provisioning/baremetal_config.go

```diff
@@ -74,57 +119,93 @@ func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1a
 	}
 	return nil
 }
 
-func validateUnmanagedConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
-		}
+func validateProvisioningInterface(iface string) error {
+	// Unfortunately there seems to be nothing you can really do to verify
+	// the ProvisioningInterface as it's not necessarily on this machine
+	// and could be almost anything.
+
+	if iface == "" {
+		return fmt.Errorf("ProvisioningInterface is required but is not set")
 	}
 	return nil
 }
 
-func validateDisabledConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
+func validateProvisioningNetworkSettings(ip string, cidr string, dhcpRange string) error {
+	// provisioningIP and networkCIDR are always set.  DHCP range is optional
+	// depending on mode.
+
+	// Verify provisioning ip and get it into net format for future tests.
+	provisioningIP := net.ParseIP(ip)
+	if provisioningIP == nil {
+		return fmt.Errorf("ProvisioningIP %q is not a valid IP", ip)
+	}
+
+	// Verify Network CIDR
+	_, provisioningCIDR, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return fmt.Errorf("%q is not a valid provisioningNetworkCIDR: %s", cidr, err)
+	}
+
+	// Ensure provisioning IP is in the network CIDR
+	if !provisioningCIDR.Contains(provisioningIP) {
+		return fmt.Errorf("%q is not part of the provisioning network CIDR", ip)
+	}
+
+	// Verify DHCP Range if required
+	if dhcpRange != "" {
+		dhcpRangeSplit := strings.Split(dhcpRange, ",")
+		if len(dhcpRangeSplit) != 2 {
+			return fmt.Errorf("%q is not a valid DHCP range.  DHCP range format: start_ip,end_ip", dhcpRange)
+		}
+		for _, ip := range dhcpRangeSplit {
+			// Ensure IP is valid
+			dhcpIP := net.ParseIP(ip)
+			if dhcpIP == nil {
+				return fmt.Errorf("The entry %q in the DHCPRange is not a valid IP", dhcpIP)
+			}
+
+			// Validate IP is in the provisioning network
+			if !provisioningCIDR.Contains(dhcpIP) {
+				return fmt.Errorf("%q is not part of the provisioning network", dhcpIP)
+			}
+		}
+
+		// Ensure provisioning IP is not in the DHCP range
+		start := net.ParseIP(dhcpRangeSplit[0])
+		end := net.ParseIP(dhcpRangeSplit[1])
+
+		if start != nil && end != nil {
```

Comment:
In theory this should never happen because we've verified the translation to IPs above.  It's just done in a for loop there to perform 2 checks on each ip.